engine/src/main/java/org/hibernate/search/backend/impl/lucene/analysis/ConcurrentlyMutableAnalyzer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.impl.lucene.analysis;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.DelegatingAnalyzerWrapper;
+import org.hibernate.search.exception.AssertionFailure;
+import org.hibernate.search.util.impl.ScopedAnalyzer;
+
+/**
+ * An Analyzer implementation which can dynamically be reconfigured. We use this as temporary solution to workaround
+ * LUCENE-6212, until a better solution is found.
+ *
+ * @author Sanne Grinovero <sanne@hibernate.org> (C) 2015 Red Hat Inc.
+ */
+public final class ConcurrentlyMutableAnalyzer extends DelegatingAnalyzerWrapper {
+
+	private final AtomicReference<ScopedAnalyzer> current = new AtomicReference<>();
+
+	public ConcurrentlyMutableAnalyzer(Analyzer initialAnalyzer) {
+		super( new ResettableReuseStrategy() );
+		if ( initialAnalyzer instanceof ScopedAnalyzer ) {
+			current.set( (ScopedAnalyzer) initialAnalyzer );
+		}
+		else {
+			current.set( new ScopedAnalyzer( initialAnalyzer ) );
+		}
+	}
+
+	@Override
+	protected Analyzer getWrappedAnalyzer(String fieldName) {
+		return current.get();
+	}
+
+	/**
+	 * Checks if the currently set Analyzer is the same as the
+	 * proposed one, in which case there is no need for
+	 * replacements or locking.
+	 * Correct concurrency control requires external locking!
+	 * @param analyzer
+	 * @return true if there is no needto replace the current Analyzer
+	 */
+	public boolean isCompatibleWith(ScopedAnalyzer analyzer) {
+		ScopedAnalyzer currentAnalyzer = current.get();
+		return currentAnalyzer.isCompositeOfSameInstances( analyzer );
+	}
+
+	/**
+	 * Correct concurrency control requires external locking!
+	 * @param analyzer
+	 */
+	public void deployNewAnalyzer(ScopedAnalyzer analyzer) {
```

Comment:
I'll use `updateAnalyzer` as in various experiments this was doing much more complex things than just the _set_ it turned into.. that might still need to change, I wouldn't want to make a promise about it being just a set